data/src/main/java/com/microsoft/azure/kusto/data/auth/MsalTokenProviderBase.java

```diff
@@ -41,13 +48,38 @@ private String determineAadAuthorityUrl(String authorityId) {
 
     @Override
     public String acquireAccessToken() throws DataServiceException, DataClientException {
+        initializeCloudInfo();
         IAuthenticationResult accessTokenResult = acquireAccessTokenSilently();
         if (accessTokenResult == null) {
             accessTokenResult = acquireNewAccessToken();
         }
         return accessTokenResult.accessToken();
     }
 
+    protected void initializeCloudInfo() throws DataServiceException, DataClientException {
+        if (cloudInfo != null) {
+            return;
+        }
+        synchronized (this) {
+            if (cloudInfo != null) {
+                return;
+            }
+
+            cloudInfo = CloudInfo.retrieveCloudInfoForCluster(clusterUrl);

```

Comment:
This can lead to race conditions.
Set it last or use a different property to determine completion and set it last

Reply:
I don't see how, the whole thing is synchronized.
Another thread will have to wait until all of the fields are initialized, and then will enter and immediately exit on `cloudInfo != null`. Did I miss something?

Reply:
first two lines are out of the sync scope
```
protected void initializeCloudInfo() throws DataServiceException, DataClientException {
        if (cloudInfo != null) {
            return;
        }
        synchronized (this) {
```

Reply:
That's on purpose, it's the Double-Checked locking pattern.
So if cloudInfo is initialized it saves us entering the synchronized block.
I also used it in python in TokenProviderBase._init_once and CloudSettings.get_cloud_info_for_cluster

Maybe it neesd a comment?

Reply:
its clear enough

Reply:
Check out the section named "The Visibility Issue"
https://mortoray.com/2012/02/28/double-checked-locking-why-doesnt-it-work-and-how-to-fix-it/

Reply:
According to this: https://stackoverflow.com/questions/3578604/how-to-solve-the-double-checked-locking-is-broken-declaration-in-java
The simplest way to solve this is to declare the field as volatile, which I now did

Reply:
There are tons of problems with double check lock 😄 
The one I refer to in particular is this:
* You use the state of cloudInfo as an indicator to skip init
* Inside the initialization block you set cloud info first
* You follow to initialize other members
* Other threads may see that cloudInfo is not null and assume that scopes and aadAuthorityUrl are also initialized 
* Hence a race condition

Reply:
Ohhhh!
Great catch

I replaced it now with a boolean `isCloudInfoInit`, that should prevent the race condition

Reply:
I think Assaf current implemenation i good although in the above stackoverflow it is discussed that the alternative to that is to use singleton which this thing really is  - then maybe we should do it like that 
If you ask me i don't think this is race is relevant here - we used to have the same methodology with AAD auth before Msal with no problems .

Reply:
It took me until Yochai's comment to see this
The methodology was fine, the problem was with `cloudInfo` as the sentinel variable.
We initialize other things after cloudInfo, so a thread that will exit in the second line of the function with a non-null cloudInfo might not have working `scopes` or  `aadAuthorityUrl`

That's why I changed it to a boolean that only gets set at the end of the function.